tests/test_asyncio.py

```diff
@@ -0,0 +1,416 @@
+                            "create_async" if create_async else "create_sync",
+                            "write_async" if write_async else "write_sync",
+                        ]
+                    )
+                    w_testcases.append(
+                        pytest.param(
+                            create_loop,
+                            expr_async,
+                            callback_async,
+                            create_async,
+                            write_async,
+                            id=idstr,
+                        )
+                    )
+
+
+@pytest.mark.parametrize(
+    "create_loop,expr_async,callback_async,create_async,write_async", w_testcases
+)
+def test_asyncio_watcher_multi(
+    create_loop, expr_async, callback_async, create_async, write_async
+):
+    loop = create_loop()
+    a = reactive([1, 2])
+    called = 0
+    completed = 0
+
+    if expr_async:
+
+        async def _expr():
+            return len(a)
+
+    else:
+
+        def _expr():
+            return len(a)
+
+    if callback_async:
+
+        async def _callback():
+            nonlocal called, completed
+            called += 1
+            await asyncio.sleep(0)
+            completed += 1
+
+    else:
+
+        def _callback():
+            nonlocal called, completed
+            called += 1
+            completed += 1
+
+    if create_async:
+
+        async def _create():
+            return watch(_expr, _callback, sync=True)
+
+        _ = loop.run_until_complete(_create())
+    else:
+        _ = watch(_expr, _callback, sync=True)
+
+    # ASSERTS pt I
+    assert (called, completed) == (0, 0)
+    flush(loop)
+    assert (called, completed) == (0, 0)
+
+    if write_async:
+
+        async def _write():
+            a.append(3)
+
+        loop.run_until_complete(_write())
+    else:
+        a.append(3)
+
+    # ASSERTS pt II
+    if (
+        create_loop is create_plain_loop
+        and not expr_async
+        and callback_async
+        and write_async
+    ):
+        assert (called, completed) == (1, 0)
+    elif (
+        create_loop is create_plain_loop
+        and expr_async
+        and callback_async
+        and not create_async
+        and write_async
+    ):
+        assert (called, completed) == (1, 0)
+    elif create_loop is create_plain_loop and expr_async and create_async:
+        assert (called, completed) == (0, 0)
+    elif (
+        create_loop is create_eager_loop and expr_async and create_async and write_async
+    ):
+        assert (called, completed) == (0, 0)
+    else:
+        assert (called, completed) == (1, 1)
+    flush(loop)
+    if create_loop is create_plain_loop and expr_async and create_async:
+        assert (called, completed) == (0, 0)
+    elif (
+        create_loop is create_eager_loop and expr_async and create_async and write_async
+    ):
+        assert (called, completed) == (0, 0)
+    else:
+        assert (called, completed) == (1, 1)
```

Comment:
This section of the test is a little too much to handle for my brains. I just reverse engineered the `assert` statements from the outputs. I _think_ they make sense... but it's a lot to process.

Reply:
IMO you should write one test function for each combination:
```py
def test_sync_watchable_with_sync_callback(): ...
def test_sync_watchable_with_async_callback(): ...

def test_async_watchable_with_sync_callback(): ...
def test_async_watchable_with_async_callback(): ...
```
Then maybe move the setup boilerplate to helper functions as needed.

At some point, and _especially_ in tests, some degree of duplication is better than convoluted logic that is hard to follow and refactor.

Reply:
Well, yes, but there are 48 cases in total... :(

Reply:
After reviewing more carefully, I think only the case `create_loop is create_eager_loop and expr_async and create_async and write_async` is suspicious and might be a bug. I'll step through that one.

Reply:
I figured it out, no bugs! I documented all the assert cases with extensive code comments.